### PR TITLE
Alerts fault test

### PR DESCRIPTION
**Pull request by damienomurchu**
## What
https://issues.redhat.com/browse/MGDSTRM-544
Add script to check for critical alerts when chaos experiments are running.

## Why
Validate chaos experiments being run

## How
- Adds test script that polls prometheus for critical alerts firing when chaosschedules are run.
- Updates make targets to add test script

## Verification:
```
cd install/resources/testing-chaos
make all
```

### CODE REVIEW

File: install/resources/testing-chaos/scripts/fault-test.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  critical_alerts = []
+  while(experiment_complete == False):
+    print("Waiting {} seconds for next Prometheus scrape interval".format(prometheus_scrape_interval))
+    sleep(prometheus_scrape_interval)
+    new_alerts = get_critical_alerts()
+    print("Alerts scraped. Critical alerts found: {}.".format(new_alerts))
+    critical_alerts.extend(new_alerts)
+    if(len(critical_alerts) > 0):
+      return critical_alerts
+    experiment_complete = are_schedules_completed(get_schedule_names(litmus_ns))
```
Comment (damienomurchu) [inline install/resources/testing-chaos/scripts/fault-test.py:72]: Ideally should check for chaosengines that are present - will look to update the script to check for those also

File: install/resources/testing-chaos/Makefile [hunk 2]
```diff
@@ -90,6 +93,10 @@ create/chaosengine/pod-delete:
     sed -e 's|<force-delete>|$(POD_DELETE_FORCE)|g' | \
     cat | oc apply -f -
 
+.PHONY: test/chaosschedule/critical-alerts
+test/chaosschedule/critical-alerts:
+	@sh ./scripts/fault-test.py $(LITMUS_NS) $(CLUSTER_PROMETHEUS_NS) $(PROMETHEUS_SCRAPE_INTERVAL)
```
Comment (jhellar) [inline install/resources/testing-chaos/Makefile:98]:
```suggestion
	@python3 ./scripts/fault-test.py $(LITMUS_NS) $(CLUSTER_PROMETHEUS_NS) $(PROMETHEUS_SCRAPE_INTERVAL)
```

File: install/resources/testing-chaos/scripts/fault-test.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+def get_critical_alerts() -> (list):
+  print("Fetching critical alerts...")
+  alerts = get_alerts(get_alerts_endpoint(prometheus_ns))
+  if(len(alerts) == 0):
+    return []
+  return filter_critical_alerts(alerts)
+
+def watch_alerts() -> (list):
+  experiment_complete = False
+  critical_alerts = []
```
Comment (jhellar) [inline install/resources/testing-chaos/scripts/fault-test.py:63]: Is this var needed, when we return from this function immediately after some critical_alerts are found?

**Review by jhellar — COMMENTED** [review]
LGTM, only thing is that I had to use `python3` in Makefile, to make it run.

File: install/resources/testing-chaos/scripts/fault-test.py (hunk 3 above)
Comment (damienomurchu) [reply install/resources/testing-chaos/scripts/fault-test.py:79]: Good point. 

I'm not sure its best to return early once any critical alerts are found. We really want to gather as much data as we can from the fault test runs I think, so I've updated the function that to remove the early return so polling for alerts will continue as long as the chaosengines are not completed.

File: install/resources/testing-chaos/scripts/fault-test.py (hunk 1 above)
Comment (damienomurchu) [reply install/resources/testing-chaos/scripts/fault-test.py]: updated now to check for chaosengines (which will implicitly include all scheduled chaos experiments also)

**Comment by damienomurchu** [comment]
Thanks for the feedback @jhellar I've responded to it in the comments above and updated PR, if you can re-review and approve if you're happy/

I've also re-run this locally against an OSD cluster and should behave as expected;

```sh
❯ make create/chaosschedule/pod-delete-zookeepers 
create pod-delete chaosschedule
chaosschedule.litmuschaos.io/pod-delete-kafka-zookeepers created

❯ make create/chaosengine/pod-delete 
create pod-delete chaosengine
chaosengine.litmuschaos.io/pod-delete created

❯ make test/chaosschedule/critical-alerts        
Waiting 30 seconds for next Prometheus scrape interval
Fetching critical alerts...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    41  100    41    0     0    506      0 --:--:-- --:--:-- --:--:--   512
Alerts scraped. Critical alerts found: [].
Waiting 30 seconds for next Prometheus scrape interval
Fetching critical alerts...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    41  100    41    0     0    341      0 --:--:-- --:--:-- --:--:--   341
Alerts scraped. Critical alerts found: [].
Waiting 30 seconds for next Prometheus scrape interval
Fetching critical alerts...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    41  100    41    0     0    173      0 --:--:-- --:--:-- --:--:--   172
Alerts scraped. Critical alerts found: [].
Waiting 30 seconds for next Prometheus scrape interval
Fetching critical alerts...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    41  100    41    0     0   1051      0 --:--:-- --:--:-- --:--:--  1051
Alerts scraped. Critical alerts found: [].
PASS: No critical alerts found
```

File: install/resources/testing-chaos/Makefile (hunk 2 above)
Comment (damienomurchu) [reply install/resources/testing-chaos/Makefile]: updated now

File: install/resources/testing-chaos/scripts/fault-test.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+def filter_critical_alerts(alerts:list) -> (list):
+  critical_alerts = []
+  for alert in alerts:
+    if(alert["labels"]["severity"] == "critical"):
+      critical_alerts.append(alert)
+  return critical_alerts  
+
+def get_all_alerts(prometheus_alerts_endpoint:str) -> (list): 
+  prometheus_json = json.loads(get_cmd_output("curl " + prometheus_alerts_endpoint))
```
Comment (jhellar) [inline install/resources/testing-chaos/scripts/fault-test.py:67]:
```suggestion
  prometheus_json = json.loads(get_cmd_output("curl -s " + prometheus_alerts_endpoint))
```

**Review by jhellar — APPROVED** [review]
Last suggestion about using `-s` for `curl`, otherwise LGTM.

**Pull request merged**